New Tooltips

alexkim205 opened the pull request:
## Changes

Implements new insight tooltips.

Closes https://github.com/PostHog/posthog/issues/7370

- [ ] Fixed bug where hover over dotted line showed duplicate/no tooltip items.
- [ ] Flip pill order to show breakdowns then time (i.e., Chrome, Current)
- [ ] Use `midEllipsis for long breakdown values (i.e., current url)`. Showing the first x characters of `http://localhost:8000/...` for all series is indifferentiable.
- [ ] Tightly coupled FF `insight-tooltips` to `line-graph-v2` to significantly simplify code. Removing `insight-tooltips`
- [ ] Various other bug fixes.
- [ ] Deprecate `chartLabel` in LineGraph.

### Caveats

To handle situations with multiple series (as @clarkus explained in detail [here](https://github.com/PostHog/posthog/issues/7370#issuecomment-996086754)), I've implemented a stopgap solution that truncates rows to 8 and columns to 5 to prevent a single tooltip from getting too large. In these cases, the situation is explained in copy below the table. 

Everything seems to work which is good, but not everything is polished yet (see below scenarios). Would love to get some stylistic recommendations here.

### Screenshots of different combinations

Multiple series, nothing else

<img width="1105" alt="Screen Shot 2022-01-04 at 4 25 15 PM" src="https://user-images.githubusercontent.com/13460330/148141702-262f2332-ecdf-4eea-ae88-2dc6ef3d8fee.png">

Multiple series, only compare

<img width="1098" alt="Screen Shot 2022-01-04 at 4 24 39 PM" src="https://user-images.githubusercontent.com/13460330/148141675-02040587-7737-4ad6-b550-5dbb64adf158.png">

Multiple series, cohorts + compare

<img width="1097" alt="Screen Shot 2022-01-04 at 4 35 17 PM" src="https://user-images.githubusercontent.com/13460330/148142403-275cf286-c09d-44c3-8ef8-caaced4de059.png">

Multiple series, breakdowns + compare

_ignore how there's no previous series. bug with local dev CH_

<img width="1244" alt="Screen Shot 2022-01-04 at 4 49 18 PM" src="https://user-images.githubusercontent.com/13460330/148143401-3b50e996-bce0-4c7a-b137-3e71d2a1abdf.png">

Multiple series, breakdowns ($current_url) -> many values truncated

<img width="1098" alt="Screen Shot 2022-01-04 at 4 39 45 PM" src="https://user-images.githubusercontent.com/13460330/148142722-319dab8b-09cd-4cac-9573-a35437c82305.png">

Single series, nothing else

<img width="1100" alt="Screen Shot 2022-01-04 at 4 40 13 PM" src="https://user-images.githubusercontent.com/13460330/148142753-2da8db7c-1886-46cb-aa00-5ef13bac56db.png">

Single series, only compare

<img width="1104" alt="Screen Shot 2022-01-04 at 4 40 35 PM" src="https://user-images.githubusercontent.com/13460330/148142784-27c19d7b-80d0-4a9e-b290-85d59dd39494.png">

Single series, only breakdown

<img width="1101" alt="Screen Shot 2022-01-04 at 4 41 06 PM" src="https://user-images.githubusercontent.com/13460330/148142824-10ed4d6c-f4f3-403d-adad-fcba4140d426.png">

Value bar graph (reuses `LineGraph.tsx`)

<img width="1102" alt="Screen Shot 2022-01-04 at 4 42 02 PM" src="https://user-images.githubusercontent.com/13460330/148142896-9eb7b7bc-90a7-4d69-9df8-7d8e4f05ed71.png">

Time bar graph (also reuses `LineGraph.tsx`)

<img width="1105" alt="Screen Shot 2022-01-04 at 4 42 50 PM" src="https://user-images.githubusercontent.com/13460330/148142955-7ef116f9-2bbb-4c10-9270-da23602a2893.png">

Legacy Line Graph (FF off)

<img width="1240" alt="Screen Shot 2022-01-04 at 4 47 52 PM" src="https://user-images.githubusercontent.com/13460330/148143309-9b290db1-b723-4183-8f4a-4645eaea648e.png">

Stickiness (x days label)

<img width="1242" alt="Screen Shot 2022-01-04 at 5 09 24 PM" src="https://user-images.githubusercontent.com/13460330/148144875-9c24cf83-f445-4d33-a23c-d5edf909941a.png">

## How did you test this code?

Regression tests pass and manual QA of all combinations.

### Out of scope (to be tackled later)

- [ ] Lifecycle insight
- [ ] New Retention Line Graph
- [ ] Fixing existing bugs (except the minor ones)
- [ ] Funnel insight tooltips (still using legacy)

alexkim205 committed:
implement first iteration of tooltips

alexkim205 committed:
add columns of data

alexkim205 committed:
Merge remote-tracking branch 'origin' into feat/new-insight-tooltips-v1

clarkus reviewed (commented):
This is looking like a really good start. I am still reviewing, but my immediate feedback is that legends should be hidden by default once we have these new tooltips. It really constrains the chart and is less valuable when tooltips are in play.

alexkim205 committed:
feat: rest of it

alexkim205 committed:
Merge remote-tracking branch 'origin' into feat/new-insight-tooltips-v1

alexkim205 commented:
@clarkus Thanks for the quick lookaround. Yup these insight legends should be hidden by default already. I wish I hid it in the screenshots above though 😓

alexkim205 committed:
typing

paolodamico commented on frontend/src/lib/components/InsightLegend/InsightLegend.tsx:
```diff
@@ -67,6 +67,7 @@ export function InsightLegend(): JSX.Element {
                                             item.breakdown_value === '' ? 'None' : item.breakdown_value?.toString()
                                         }
                                         compareValue={filters.compare ? formatCompareLabel(item) : undefined}
+                                        pillMidEllipsis={item?.filter?.breakdown === '$current_url'}
```
Curious why we only do this for `$current_url`, in any case adding a code comment could be useful.

alexkim205 replied:
It would be more robust if we could define a set of breakdown values that would benefit from a mid ellipsis truncation. For now, I thought $current_url was the biggest offender of taking up too much space but not having differentiating enough text in the beginning or end of the string. But I agree I'll add a todo comment here for now.

paolodamico commented on frontend/src/scenes/insights/InsightTooltip/InsightTooltip.tsx:
```diff
@@ -1,50 +1,160 @@
+export function InsightTooltip({ date, seriesData = [], useAltTitle }: InsightTooltipProps): JSX.Element {
+    /*
+     How to format each row's title depending on whether breakdown or compare exists
+     Breakdown  | Y | Y | N | N
```
Unsure what this comment is describing.

paolodamico commented on frontend/src/scenes/insights/InsightTooltip/insightTooltipUtils.tsx:
```diff
@@ -0,0 +1,81 @@
+import { capitalizeFirstLetter, midEllipsis, pluralize } from 'lib/utils'
+
+export interface SeriesDatum {
+    id: number
```
Worth documenting what each of the different IDs are

paolodamico commented on frontend/src/scenes/insights/InsightTooltip/insightTooltipUtils.tsx:
```diff
@@ -0,0 +1,81 @@
+    count: number
+}
+
+export interface InvertedSeriesDatum {
```
Worth documenting when do we use this

paolodamico reviewed (changes requested):
Looking amazing! Some UX comments,
1. Unsure if we can, but interacting with the tooltip would be amazing. The "Click to view users" is confusing as you can't actually click it.
2. When you have multiple series, I think we should rely on the series glyph
    <img width="1029" alt="" src="https://user-images.githubusercontent.com/5864173/148275906-9c54bc58-8004-4ae3-8611-a4c64be337bd.png">
3. Seems unintuitive that the tooltip in the bar value chart doesn't display the breakdown value.
    <img width="423" alt="" src="https://user-images.githubusercontent.com/5864173/148276038-73ac0354-cf03-4032-b9c1-284e76b5539d.png">
4. This [retention graph](http://localhost:8000/insights/S26x4a3m/edit?insight=RETENTION&target_entity=%7B%22id%22%3A%22%24pageview%22%2C%22name%22%3A%22%24pageview%22%2C%22type%22%3A%22events%22%7D&returning_entity=%7B%22id%22%3A%22%24pageview%22%2C%22type%22%3A%22events%22%2C%22name%22%3A%22%24pageview%22%7D&period=Day&retention_type=retention_first_time&display=ActionsTable&properties=%5B%5D) showed "no entries" for some reason instead of 0.
    <img width="407" alt="" src="https://user-images.githubusercontent.com/5864173/148276171-3dd9f5c2-9885-4d4d-812e-6a8960f8fb7e.png">
5. Formulas don't seem to be generating a correct tooltip.
    <img width="323" alt="" src="https://user-images.githubusercontent.com/5864173/148276319-d5097bf0-093b-4496-9786-65593dbd939b.png">
6. Unsure if this is related to this PR but when I tried to create a graph broken down by `$current_url` I got the following CH error.
<img width="534" alt="" src="https://user-images.githubusercontent.com/5864173/148276521-ddb28dfb-e15c-4048-b5a3-5553ae2a5e10.png">

alexkim205 commented:
Thanks for finding these bugs!
 
> 1. Unsure if we can, but interacting with the tooltip would be amazing. The "Click to view users" is confusing as you can't actually click it.

Unfortunately this is a hard constraint set by ChartJs's limitations. The position of the cursor will always take priority and theres no way of disabling interactivity with canvas elements via just js.

Fixed 2-5 and addressed inline comments. I'm unable to reproduce #6 but I suspect this is an already existing bug since it's crashing on query time.

A few more changes:

- Dropped row color border from tooltip when compare/breakdown is applied. These are inaccurate as colors are series specific, not breakdown.

alexkim205 committed:
various fixes

Twixes commented on frontend/src/scenes/insights/InsightTooltip/InsightTooltip.scss:
```diff
@@ -0,0 +1,89 @@
+            overflow: hidden;
+        }
+        .LemonTable__content > table > thead {
+            text-transform: none;
```
In this case I also suggest overriding `letter-spacing`, as it's normally increased in the header to make the all-caps titles more relaxed.
```suggestion
            text-transform: none;
            letter-spacing: normal;
```
BTW: you may consider making this a prop on `LemonTable`, something like `uppercaseHeader` (which would be `true` by default) – I could see the alternative style being useful outside of just the tooltips

alexkim205 replied:
moved into a prop

Twixes commented on frontend/src/scenes/insights/LineGraph/LineGraph.tsx:
```diff
@@ -310,40 +314,27 @@ export function LineGraph(props: LineGraphProps): JSX.Element {
-                                            inspectPersonsLabel={onClick && showPersonsModal}
-                                        />
-                                    )}
+                                    <InsightTooltip
```
On a related (but loosely) note, there's still a problem where navigating away from the scene while the insight tooltip is shown results in the insight tooltip being totally stuck:
<img width="284" alt="Screen Shot 2022-01-06 at 12 04 28" src="https://user-images.githubusercontent.com/4550621/148375209-5921d1e3-e038-4852-9e04-240babfe54f1.png">
We should somehow stop rendering it once the `LineGraph` component is not shown anymore.

alexkim205 replied:
This should be fixed now

Twixes commented on frontend/src/scenes/insights/LineGraph/lineGraphLogic.ts:
```diff
@@ -0,0 +1,54 @@
+            () => [],
+            () =>
+                (tooltipDataPoints: TooltipItem<any>[], filterFn: (s: SeriesDatum) => boolean): SeriesDatum[] => {
+                    console.log('DATAPOINTS,', tooltipDataPoints)
```
Leftover `console.log`

alexkim205 replied:
arghh nice catch

Twixes commented on frontend/src/scenes/insights/LineGraph/lineGraphLogic.ts:
```diff
@@ -0,0 +1,54 @@
+                    return tooltipDataPoints
+                        ?.map((dp, idx) => {
+                            const pointDataset = (dp?.dataset ?? {}) as GraphDataset
+                            console.log('POINT DATASET', pointDataset)
```
`console.log`

Twixes commented on frontend/src/scenes/insights/LineGraph/lineGraphLogic.ts:
```diff
@@ -0,0 +1,54 @@
+export const lineGraphLogic = kea<lineGraphLogicType>({
+    path: ['scenes', 'insights', 'LineGraph', 'lineGraphLogic'],
+    selectors: {
+        createTooltipData: [
+            () => [],
+            () =>
+                (tooltipDataPoints: TooltipItem<any>[], filterFn: (s: SeriesDatum) => boolean): SeriesDatum[] => {
```
I don't quite see why this is a Kea logic instead of just a standalone `createTooltipData` util?

alexkim205 replied:
There's no real reason, but I wanted to start moving away state from the linegraph component and into its own separate logic. This is my attempt at chipping away at that tech debt, but perhaps a todo comment here will also encourage people to add things here too

Twixes commented on frontend/src/scenes/insights/InsightTooltip/InsightTooltip.scss:
```diff
@@ -0,0 +1,89 @@
+        background-color: $bg_light;
+
+        .table-subtext-click-to-inspect {
+            margin-top: $default_spacing / 4;
```
This `margin-top` causes the tip to be a bit off-center vertically in the basic case. I understand that it's needed when the "not all series are displayed" message is shown, but the margin should only apply to the class if it's `:not(:first-child)`.
<img width="196" alt="Screen Shot 2022-01-06 at 12 28 07" src="https://user-images.githubusercontent.com/4550621/148376214-601eba92-0e97-4869-8248-4822f7c893d2.png">

alexkim205 replied:
nice catch!

Twixes commented on frontend/src/scenes/insights/InsightTooltip/InsightTooltip.scss:
```diff
@@ -0,0 +1,89 @@
+
+        .table-subtext-click-to-inspect {
+            margin-top: $default_spacing / 4;
+            color: $primary;
```
The blue text color of "Click to view users" makes me think it's a clickable link, while it's just a suggestion for the user. Slightly frustrating because it makes the tooltip look interactive when it isn't at all. I propose using gray similar to what the previous tooltip iteration.
<img width="196" alt="Screen Shot 2022-01-06 at 12 28 07" src="https://user-images.githubusercontent.com/4550621/148376336-c82dd790-4557-4557-87e0-ec279ff80320.png">

Twixes commented on frontend/src/scenes/insights/InsightTooltip/InsightTooltip.scss:
```diff
@@ -0,0 +1,89 @@
+    .table-subtext {
+        font-size: 0.75rem;
+        text-align: center;
+        padding: $default_spacing / 2 $default_spacing;
+        color: $text_muted;
+        background-color: $bg_light;
```
Nit: let's minimize usage of preprocessor variables. See https://github.com/PostHog/posthog/pull/7726#discussion_r770674509.
There's also IMO an additional problem with `$default_spacing` in particular, in that it's quite opaque what it is, using it requires additional math like the division here AND there is little utility in making it a variable (it's not like we're going to EVER change it across the board – in reality the spacing must be tuned everywhere anyway). The `rem` unit (which is `16px` in PostHog too) seems like a better way of specifying sizes.

alexkim205 replied:
thanks, thats good to know!

Twixes commented on frontend/src/scenes/insights/InsightTooltip/insightTooltipUtils.tsx:
```diff
@@ -0,0 +1,109 @@
+        }
+        return tooltipAltTitle
+    }
+    return getFormattedDate(date)
```
This works for most cases BUT when grouping by hour it doesn't show the relevant hour – I think it really should.

Twixes reviewed (dismissed):
This looks really nice! A few more polishes and this will be great.

paolodamico reviewed (approved):
Thanks @alexkim205! lgtm **UX-wise**, some additional comments (non-blocking),

- I would show the series glyph whenever is multiple series. With the latest change we only show it when comparing to previous.
- The pattern across the app has been to show the series glyph only if there's more than one series. I think we should follow this (even if you're comparing previous).

- For the retention tooltip I would do some special handling and add the word "Cohort" in front of each (even "Day cohort" or "Week cohort") to clarify. Otherwise seeing a date there seems a bit confusing.
    <img width="324" alt="" src="https://user-images.githubusercontent.com/5864173/148431219-0d9be5ef-c92a-480f-bd80-c7332a27fb3e.png">

alexkim205 commented:
Thanks for the quick review!

> * I would show the series glyph whenever is multiple series. With the latest change we only show it when comparing to previous.

The series glyphs should already show in both breakdown and compare situations. Do you mean when there are no breakdowns/compare applied? For now the series are differentiated by color but I can add a glyph in there too. This is what it looks like now for multi series no breakdown/no cohorts tooltips:

<img width="327" alt="Screen Shot 2022-01-06 at 11 41 46 AM" src="https://user-images.githubusercontent.com/13460330/148441665-aa194101-2961-4dfe-9d3c-b4413046a7ec.png">


> * The pattern across the app has been to show the series glyph only if there's more than one series. I think we should follow this (even if you're comparing previous).

Hm, not sure what you mean here, could you clarify? Compare previous already shows the series glyph when the series are itemized as columns for compare.

<img width="237" alt="Screen Shot 2022-01-06 at 11 03 58 AM" src="https://user-images.githubusercontent.com/13460330/148437122-223bc72d-50a0-4a3e-8378-1ed86dcbb851.png">

> * For the retention tooltip I would do some special handling and add the word "Cohort" in front of each (even "Day cohort" or "Week cohort") to clarify. Otherwise seeing a date there seems a bit confusing.

👍  good idea. I'm going to only put Cohort for now, since X cohort crowds up space and time unit can be inferred from the x axis

alexkim205 committed:
fix tests and address feedback

alexkim205 committed:
Merge remote-tracking branch 'origin' into feat/new-insight-tooltips-v1

alexkim205 committed:
typing

alexkim205 committed:
Merge remote-tracking branch 'origin' into feat/new-insight-tooltips-v1

alexkim205 committed:
fix tests

paolodamico reviewed (approved):
lgtm @alexkim205! I meant that when there's only one series, don't show the glyph. Though it looks like we started showing it in the query builder all the time.

alexkim205 commented:
👍 I'll keep the glyphs in there for now so that it's at least consistent with the filters.
